quinn-proto/src/connection/streams/mod.rs

```diff
@@ -335,60 +335,43 @@ impl<'a> SendStream<'a> {
     }
 }
 
-fn push_pending(pending: &mut BinaryHeap<PendingLevel>, id: StreamId, priority: i32) {
-    for level in pending.iter() {
-        if priority == level.priority {
-            level.queue.borrow_mut().push_back(id);
-            return;
-        }
-    }
-
-    // If there is only a single level and it's empty, repurpose it for the
-    // required priority
-    if pending.len() == 1 {
-        if let Some(mut first) = pending.peek_mut() {
-            let mut queue = first.queue.borrow_mut();
-            if queue.is_empty() {
-                queue.push_back(id);
-                drop(queue);
-                first.priority = priority;
-                return;
-            }
-        }
-    }
-
-    let mut queue = VecDeque::new();
-    queue.push_back(id);
-    pending.push(PendingLevel {
-        queue: RefCell::new(queue),
-        priority,
-    });
+/// Push a pending stream ID with the given priority, queued after all preexisting streams for the priority
+fn push_pending(pending: &mut BTreeMap<PendingPriority, StreamId>, id: StreamId, priority: i32) {
+    // Get all preexisting streams of this priority
+    // We only actually need the lowest recency value, so this could be implemented more tersely with
+    // `BTreeMap::lower_bound`, but as of writing that API is unfortunately still nightly only
+    let mut range = pending.range((
+        Bound::Included(PendingPriority {
+            priority,
+            recency: 0,
+        }),
+        Bound::Included(PendingPriority {
+            priority,
+            recency: u64::MAX,
+        }),
+    ));
+
+    pending.insert(
+        PendingPriority {
+            priority,
+            recency: range.next().map(|(p, _)| p.recency - 1).unwrap_or(u64::MAX),
```

Comment:
Could `p.recency - 1` wrap? If not, why not? Maybe comment that this is unlikely to wrap because `recency` gets initialized to `u64::MAX`?